Autoloader: refactor functions.php

kbrown9 opened the pull request:
This PR is based on the `feature/update_autoloader` feature branch.

#### Changes proposed in this Pull Request:
* Refactor the functions in `functions.php` into the following classes:
  * `Autoloader_Handler` - used to find the latest autoloader.
  * `Classes_Handler` - used to set the class versions in the classmap
  * `Files_Handler` - used to set the file versions in the filemap.
  * `Plugins_Handler` - used to get information about the current plugin and the active plugins.

Note: In the next PR, a new class for determining which package version to use will be added. This class will be used by `Autoloader_Handler`, `Classes_Handler`, and `Files_Handler` to select which version of the autoloader, classes, and files to use.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This changes the existing Autoloader package.

#### Testing instructions:
* Test the branch with plugins that use the autoloader and ensure that no errors are generated.

#### Proposed changelog entry for your changes:
* n/a

kbrown9 committed:
Autoloader: refactor functions.php into classes

kbrown9 committed:
Autoloader: update unit tests

Update the unit tests to use the new Classes_Handler and Files_Handler classes.

kbrown9 commented on packages/autoloader/src/class-plugins-handler.php:
```diff
@@ -0,0 +1,125 @@
+		$current_plugin = $this->get_current_plugin();
+
+		return in_array( $current_plugin, $active_plugins, true );
+	}
```
I think we might want to return True if the plugin is active or if it's a known activating plugin (see `get_activating_plugins()` ). `is_current_plugin_active()` is used in `set_up_autoloader()` to determine if the autoloader globals should be reset. If we already know that the plugin is activating, the globals shouldn't be reset.

zinigor replied:
If I understand your idea here correctly, you mean that we should look for the current plugin in the following list:
```
	$active_plugins = $this->get_active_plugins();
```

Right now we're doing this instead:
```
	$active_plugins = (array) get_option( 'active_plugins', array() );
```

If we use `get_active_plugins`, we will return true in `is_current_plugin_active` every time that we can detect the plugin activation, so the autoloader globals reset will be reserved only for the case when we can't detect its activation, right?

If that's so, I think that should work fine.

kbrown9 replied:
Yes, that's exactly what I was thinking. I made this change in commit 612e84f.

kbrown9 replied:
I forgot that `get_active_plugins()` added the current plugin to the active plugins list if it wasn't already there. So, I couldn't use that method to determine whether to reset the autoloader.

I pushed a commit that refactors the `Plugins_Handler` class a bit. With the changes in that commit, the autoloader will only reset when the plugin was unknown (i.e. not an active plugin and not an already known activating plugin).

kbrown9 commented:
A couple of comments about this PR:

 * The E2E tests are failing at the plugin activation, and I'm not sure why. I'm not having any problems activating Jetpack using WP-CLI on my local test environment.

 * The scenario below can cause problems. I haven't worked on a solution yet.
   * A plugin is activating using a method that doesn't use a request, like WP-CLI. This means that the plugin is not identified as an 'activating plugin' in `Plugins_Handler::get_activating_plugins().`
   * Another plugin with an autoloader is already active on the site.
   * The autoloader for the activating plugin is an earlier version than the autoloader for the active plugin. When the autoloader is reset by the activating plugin, the latest autoloader (from the active plugin) is selected to run.
   * The latest autoloader doesn't know that the activating plugin exists, so it will not check the classmap and filemap files for that plugin.

kbrown9 committed:
Autoloader: update is_current_plugin_active()

Update is_current_plugin_active() to check both active and activating
plugins when determining whether the current plugin is active. The
active and activating plugins can be obtained using the
Plugins_Handler::get_active_plugins() method.

zinigor committed:
Added newly generated files to gitignore.

kbrown9 committed:
Autoloader: improve Plugins_Handler

Improve the Plugins_Handler class by:
 - Giving some methods more descriptive names.
 - Adding a $jetpack_autoloader_activating_plugins global variable that
   contains the plugins that are activating via a non-request method.
 - Refactoring a few methods to improve clarity.

kbrown9 commented:
> * The E2E tests are failing at the plugin activation, and I'm not sure why. I'm not having any problems activating Jetpack using WP-CLI on my local test environment.

The E2E tests are fine now.

> * The scenario below can cause problems. I haven't worked on a solution yet.
>   
>   * A plugin is activating using a method that doesn't use a request, like WP-CLI. This means that the plugin is not identified as an 'activating plugin' in `Plugins_Handler::get_activating_plugins().`
>   * Another plugin with an autoloader is already active on the site.
>   * The autoloader for the activating plugin is an earlier version than the autoloader for the active plugin. When the autoloader is reset by the activating plugin, the latest autoloader (from the active plugin) is selected to run.
>   * The latest autoloader doesn't know that the activating plugin exists, so it will not check the classmap and filemap files for that plugin.

The changes in commit df4d569 should resolve this issue. Each time a previously unknown plugin is detected, that plugin is added to a global array, `$jetpack_autoloader_activating_plugins`. Then, the latest autoloader, regardless of which plugin it's associated with, will have a complete list of all known activating plugins.

kbrown9 committed:
Autoloader: update the autoloader chain after autoloader reset

Add a new function, Autoloader_Handler::update_autoloader_chain(), which
handles updates of the autoloader chain as follows:
 - Registers the latest autoloader function.
 - Moves the v1 autoloader function to the end of the chain.
 - Removes any other v2 autoloader functions. This is needed for situations where
   a plugin activates using a non-request method, so the activating plugin was not
   known when the autoloader first ran. If the activating plugin has an autoloader
   with a later version, that autoloader will be registered and the previously
   registered autoloader function should be removed.

Finally, use the Autoloader_Handler::update_autoloader_chain() method in
set_up_autoloader().

kbrown9 committed:
Autoloader: update update_autoloader_chain()

In Autoloader_Handler::update_autoloader_chain(), check whether the autoloader
object is a string. If not, just continue. We're only looking for Jetpack
autoloader functions, and they're registered as strings.

zinigor reviewed (approved):
This seems to be working very well! Are you planning to add anything else, or can we merge?
